popgen/plot.go

```diff
@@ -8,139 +8,189 @@ import (
 
 	//a short header is written to the file
-	fmt.Fprintf(out, "Frequency\tProbability\n")
+	_, err = fmt.Fprintf(out, "Frequency\tProbability\n")
+	exception.PanicOnErr(err)
 	//for each possible allele frequency for a segregating site, we calculate the probability to the file
 	for i := 1; i < n; i++ {
-		fmt.Fprintf(out, "%v\t%e\n", i, AlleleFrequencyProbabilityAncestralAscertainment(alpha, i, n, 1, binomCache, integralError))
+		_, err = fmt.Fprintf(out, "%v\t%e\n", i, AlleleFrequencyProbabilityAncestralAscertainment(alpha, i, n, 1, binomCache, integralError))
+		exception.PanicOnErr(err)
 	}
 
-	err := out.Close()
+	err = out.Close()
 	exception.PanicOnErr(err)
 }
 
 func PlotAfsPmfDerived(alpha float64, n int, outFile string, integralError float64) {
+	var err error
 	out := fileio.EasyCreate(outFile)
 
 	allN := []int{n}
 	binomCache := BuildBinomCache(allN)
 	//DEBUG:fmt.Printf("Len BinomCache:%v. Entry at index n is of length: %v.\n", len(binomCache), len(binomCache[n]))
 
 	//a short header is written to the file
-	fmt.Fprintf(out, "Frequency\tProbability\n")
+	_, err = fmt.Fprintf(out, "Frequency\tProbability\n")
+	exception.PanicOnErr(err)
 	//for each possible allele frequency for a segregating site, we calculate the probability to the file
 	for i := 1; i < n; i++ {
-		fmt.Fprintf(out, "%v\t%e\n", i, AlleleFrequencyProbabilityDerivedAscertainment(alpha, i, n, 1, binomCache, integralError))
+		_, err = fmt.Fprintf(out, "%v\t%e\n", i, AlleleFrequencyProbabilityDerivedAscertainment(alpha, i, n, 1, binomCache, integralError))
+		exception.PanicOnErr(err)
 	}
 
-	err := out.Close()
+	err = out.Close()
 	exception.PanicOnErr(err)
 }
 
 //PlotAfsPmf writes the allele frequency probability mass function (AlleleFrequencyProbability) to an output file for downstream visualization.
 func PlotAfsPmf(alpha float64, n int, outFile string, integralError float64) {
+	var err error
 	out := fileio.EasyCreate(outFile)
 
 	allN := []int{n}
 	binomCache := BuildBinomCache(allN)
 	//DEBUG:fmt.Printf("Len BinomCache:%v. Entry at index n is of length: %v.\n", len(binomCache), len(binomCache[n]))
 
 	//a short header is written to the file
-	fmt.Fprintf(out, "Frequency\tProbability\n")
+	_, err = fmt.Fprintf(out, "Frequency\tProbability\n")
 	//for each possible allele frequency for a segregating site, we calculate the probability to the file
 	for i := 1; i < n; i++ {
-		fmt.Fprintf(out, "%v\t%e\n", i, AlleleFrequencyProbability(i, n, alpha, binomCache, integralError))
+		_, err = fmt.Fprintf(out, "%v\t%e\n", i, AlleleFrequencyProbability(i, n, alpha, binomCache, integralError))
+	}
+
+	err = out.Close()
+	exception.PanicOnErr(err)
+}
+
+//PlotAfsLikelihood
+func PlotAfsLikelihood(afs Afs, outFile string, leftBound float64, rightBound float64, numPoints int, integralError float64, divergenceAscertainment bool, D int) {
+	var err error
+	var alpha float64
+	out := fileio.EasyCreate(outFile)
+
+	allN := findAllN(afs)
+	binomMap := BuildBinomCache(allN)
```

Comment:
Where is this BuildBinomCache function coming from? Internal to golang?

Reply:
This is a part of the MCMC code. When we calculate the likelihood function for a particular allele frequency spectrum, we often have to calculate binomial probabilities from large numbers and use those values over and over again. This helper function pre-calculates all possible binomials we will need for a particular allele frequency spectrum and saves them to a cache as a runtime optimization